controllers/kubeadmbootstrapconfig_controller_test.go

```diff
@@ -0,0 +1,159 @@
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	kubeadmv1alpha1 "sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm/api/v1alpha1"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type log struct{}
+
+func (l *log) Info(msg string, keysAndValues ...interface{})             {}
+func (l *log) Enabled() bool                                             { return false }
+func (l *log) Error(err error, msg string, keysAndValues ...interface{}) {}
+func (l *log) V(level int) logr.InfoLogger                               { return l }
+func (l *log) WithValues(keysAndValues ...interface{}) logr.Logger       { return l }
+func (l *log) WithName(name string) logr.Logger                          { return l }
+
+type myClient struct {
+	db map[string]runtime.Object
+}
+
+// Very basic implementation of copying pointers of interfaces around, stolen from controller runtime
+func (c *myClient) Get(ctx context.Context, key client.ObjectKey, out runtime.Object) error {
+	obj, ok := c.db[key.String()]
+	if !ok {
+		return errors.New("object not found")
+	}
+	obj = obj.(runtime.Object).DeepCopyObject()
+	outVal := reflect.ValueOf(out)
+	objVal := reflect.ValueOf(obj)
+	reflect.Indirect(outVal).Set(reflect.Indirect(objVal))
+	return nil
+}
+func (c *myClient) List(ctx context.Context, list runtime.Object, opts ...client.ListOptionFunc) error {
+	return nil
+}
+func (c *myClient) Create(ctx context.Context, obj runtime.Object, opts ...client.CreateOptionFunc) error {
+	return nil
+}
+func (c *myClient) Delete(ctx context.Context, obj runtime.Object, opts ...client.DeleteOptionFunc) error {
+	return nil
+}
+func (c *myClient) Update(ctx context.Context, obj runtime.Object, opts ...client.UpdateOptionFunc) error {
+	return nil
+}
+func (c *myClient) Patch(ctx context.Context, obj runtime.Object, patch client.Patch, opts ...client.PatchOptionFunc) error {
+	return nil
+}
+func (c *myClient) Status() client.StatusWriter {
+	return c
+}
+
+func TestSuccessfulReconcileShouldNotRequeue(t *testing.T) {
+	machineKind := v1alpha2.SchemeGroupVersion.WithKind("Machine").String()
+
+	myclient := &myClient{
+		db: map[string]runtime.Object{
+			"ns/cfg": &kubeadmv1alpha1.KubeadmBootstrapConfig{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "ns",
+					Name:      "cfg",
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							Kind: machineKind,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	k := &KubeadmBootstrapConfigReconciler{
+		Log:    &log{},
+		Client: myclient,
+	}
+
+	request := ctrl.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: "ns",
+			Name:      "cfg",
+		},
+	}
+	result, err := k.Reconcile(request)
+	if err != nil {
+		t.Fatal(fmt.Sprintf("%+v", err))
```

Comment:
do this error need a prefix e.g.:
`t.Fatalf("some message: %+v", err)`
[1]

Reply:
It doesn't strictly need one and since the err is printing a stack trace of where the error comes from I didn't feel it was that important, but there's no harm in adding one. I'll do that.